src/app/server/AppDelegate.h

```diff
@@ -29,11 +29,31 @@ class AppDelegate
 public:
     virtual ~AppDelegate() {}
     /**
-     * This is called on start of session establishment process
+     * This is called when the PBKDFParamRequest is received and indicates the start of the session establishment process
      */
     virtual void OnCommissioningSessionEstablishmentStarted() {}
+
+    /**
+     * This is called when the commissioning session has been established
+     */
     virtual void OnCommissioningSessionStarted() {}
-    virtual void OnCommissioningSessionStopped(CHIP_ERROR err) {}
+
+    /**
+     * This is called when there is an error in establishing a commissioning session (such as, when an invalid passcode is provided)
+     *
+     * @param err CHIP_ERROR indicating the error that occurred during session establishment
+     *
+     * @return true if the commissioning window should be closed, false otherwise
+     */
+    virtual bool OnCommissioningSessionEstablishmentError(CHIP_ERROR err) { return false; }
+
+    /**
+     * This is called in addition to OnCommissioningSessionEstablishmentError i.e. when there is an error in establishing a
+     * commissioning session AND the commissioning window is closed. The window may be closed because
+     * OnCommissioningSessionEstablishmentError returned 'true', the commissioning attempts limit was reached or listening for PASE
+     * failed.
```

Comment:
I don't think this documentation is correct as things stand.  OnCommissioningSessionStopped is called in the following cases:

1. Establishing a PASE session fails.
2. An armed failsafe expires.

If we don't go back to listening for incoming PASE as a result.

Number 2 can happen for a failsafe that is not even related to PASE at all, though that is probably just a bug we should fix: `CommissioningWindowManager` never does any `RemoveEventHandler` that I can see.

Assuming we fix that bug, we would be down to either PASE establishment failed or PASE was established fine but then the fail-safe expired (including being expired by the commissioner).

Importantly, this can be called even if establishing the PASE session happened just fine.

Reply:
I have changed the documentation to follow this statement from your comment: "Assuming we fix that bug, we would be down to either PASE establishment failed or PASE was established fine but then the fail-safe expired (including being expired by the commissioner)."

See updated diff.